File: app/src/main/java/com/carbonplayer/model/entity/ParcelableTrack.java
package com.carbonplayer.model.entity;


import android.support.v4.media.MediaMetadataCompat;

import org.parceler.Parcel;

import java.util.Objects;

/**
 * Simplified Track object for passing to/from background service
 */
@SuppressWarnings("unused")
@Parcel
public class ParcelableTrack {
    long localId;
    int position;
    String id;
    String clientId;
    String nid;
    String storeId;
    String title;
    String artist;
    String album;
    int year;
    int trackNumber;
    String genre;
    int durationMillis;
    String albumArtURL;
    String rating;
    int estimatedSize;

    ParcelableTrack() {
    }

    public ParcelableTrack(int position, String id, String title,
                           String artist, String album, int year,
                           int trackNumber, String genre,
                           int durationMillis, String albumArtURL,
                           String rating, int estimatedSize) {
        this.position = position;
        this.id = id;
        this.title = title;
        this.artist = artist;
        this.album = album;
        this.year = year;
        this.trackNumber = trackNumber;
        this.genre = genre;
        this.durationMillis = durationMillis;
        this.albumArtURL = albumArtURL;
        this.rating = rating;
        this.estimatedSize = estimatedSize;
    }

    public ParcelableTrack(Track source) {
        this.localId = source.getLocalId();
        this.id = source.getId();
        this.clientId = source.getClientId();
        this.nid = source.getNid();
        this.storeId = source.getStoreId();
        this.title = source.getTitle();
        this.artist = source.getArtist();
        this.album = Objects.requireNonNull(Objects.requireNonNull(source.getAlbums()).first()).getTitle();
        this.year = orZero(source.getYear());
        this.trackNumber = orZero(source.getTrackNumber());
        this.genre = source.getGenre();
        this.durationMillis = source.getDurationMillis();
        this.albumArtURL = Objects.requireNonNull(source.getAlbums().first()).getAlbumArtRef();
        this.rating = source.getRating();
        Integer estSize = source.getEstimatedSize();
        this.estimatedSize = estSize == null ? 0 : estSize;
    }

    @Override
    public String toString() {
        return "ParcelableTrack(id: " + this.id + ", title: " + this.title + ")";
    }

    public long getLocalId() {
        return localId;
    }

    public void setLocalId(long localId) {
        this.localId = localId;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getClientId() {
        return clientId;
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public String getNid() {
        return nid;
    }

    public void setNid(String nid) {
        this.nid = nid;
    }

    public String getStoreId() {
        return storeId;
    }

    public void setStoreId(String storeId) {
        this.storeId = storeId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getArtist() {
        return artist;
    }

    public void setArtist(String artist) {
        this.artist = artist;
    }

    public String getAlbum() {
        return album;
    }

    public void setAlbum(String album) {
        this.album = album;
    }

    public int getYear() {
        return year;
    }

    public void setYear(int year) {
        this.year = year;
    }

    public int getTrackNumber() {
        return trackNumber;
    }

    public void setTrackNumber(int trackNumber) {
        this.trackNumber = trackNumber;
    }

    public String getGenre() {
        return genre;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }

    public int getDurationMillis() {
        return durationMillis;
    }

    public void setDurationMillis(int durationMillis) {
        this.durationMillis = durationMillis;
    }

    public String getAlbumArtURL() {
        return albumArtURL;
    }

    public void setAlbumArtURL(String albumArtURL) {
        this.albumArtURL = albumArtURL;
    }

    public String getRating() {
        return rating;
    }

    public void setRating(String rating) {
        this.rating = rating;
    }

    public int getEstimatedSize() {
        return estimatedSize;
    }

    public void setEstimatedSize(int estimatedSize) {
        this.estimatedSize = estimatedSize;
    }

    public MediaMetadataCompat getMediaMetadata() {
        return new MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, getId())
            /*.putString(MusicProvider.CUSTOM_METADATA_TRACK_SOURCE, getUrl(context, realm).toString())*/
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, getAlbum())
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, getArtist())
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, getDurationMillis())
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, getAlbumArtURL())
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, getTitle())
                .putLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER, getTrackNumber())
             /*   .putLong(MediaMetadata.METADATA_KEY_NUM_TRACKS, allTracks.size())*/
                .build();
    }

    private int orZero(Integer in) {
        if (in == null) return 0;
        return in;
    }

}

File: app/src/main/java/com/carbonplayer/model/entity/ParcelableTrack.kt
package com.carbonplayer.model.entity


import android.annotation.SuppressLint
import android.os.Parcelable
import android.support.v4.media.MediaMetadataCompat
import com.carbonplayer.model.entity.base.ITrack
import com.carbonplayer.model.entity.skyjam.SkyjamTrack
import io.realm.Realm
import kotlinx.android.parcel.Parcelize

/**
 * Simplified Track object for passing to/from background service
 */
@SuppressLint("ParcelCreator")
@Parcelize
class ParcelableTrack (

        var localId: Long? = null,
        var position: Int = 0,
        override var id: String? = null,
        override var clientId: String? = null,
        override var nid: String? = null,
        override var storeId: String? = null,
        override var title: String,
        override var artist: String,
        val artistId: List<String>?,
        val artistArtRef:List<String>?,
        override var album: String,
        override var year: Int = 0,
        override var trackNumber: Int = 0,
        override var genre: String? = null,
        override var durationMillis: Int = 0,
        override var albumArtURL: String?,
        override var rating: String? = null,
        override val albumArtist: String,
        override val albumId: String,
        override val audioAd: Boolean?,
        override val beatsPerMinute: Int?,
        override val comment: String?,
        override val composer: String?,
        override val creationTimestamp: Long?,
        override val discNumber: Int,
        override val explicitType: Int?,
        override var inLibrary: Boolean,
        override val isDeleted: Boolean,
        override val lastRatingChangeTimestamp: Long?,
        override val playCount: Int?,
        override val recentTimestamp: Long?,
        override val totalDiscCount: Int?,
        override val totalTrackCount: Int?,
        override var estimatedSize: Int = 0,
        override val trackAvailableForPurchase: Boolean?,
        override val trackAvailableForSubscription: Boolean?,
        override val trackType: Int?,
        var queuePosition: Int = 0

) : ITrack, Parcelable {

    override fun parcelable(realm: Realm?) = this

    val mediaMetadata: MediaMetadataCompat
        get() = MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, id)
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, album)
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, artist)
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, durationMillis.toLong())
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, albumArtURL)
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, title)
                .putLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER, trackNumber.toLong())
                .build()

    constructor(source: Track) : this (
        source.localId, 0, source.id, source.clientId, source.nid, source.storeId,
            source.title, source.artist, source.artists?.map {it.artistId},
            source.artists?.firstOrNull()?.artistArtRef?.let { listOf(it) },
            source.album, source.year, source.trackNumber,
            source.genre, source.durationMillis, source.albums?.firstOrNull()?.albumArtRef,
            source.rating, source.albumArtist, source.albumId, source.audioAd,
            source.beatsPerMinute, source.comment, source.composer, source.creationTimestamp,
            source.discNumber, source.explicitType, source.inLibrary, source.isDeleted,
            source.lastRatingChangeTimestamp, source.playCount, source.recentTimestamp,
            source.totalDiscCount, source.totalTrackCount, source.estimatedSize ?: -1,
            source.trackAvailableForPurchase, source.trackAvailableForSubscription, source.trackType
    )

    constructor(source: SkyjamTrack) : this (
            null, 0, source.id, source.clientId, source.nid, source.storeId,
            source.title, source.artist, source.artistId, source.artistArtRef?.map {it.url},
            source.album, source.year, source.trackNumber,
            source.genre, source.durationMillis, source.albumArtRef?.firstOrNull()?.url,
            source.rating, source.albumArtist, source.albumId, source.audioAd,
            source.beatsPerMinute, source.comment, source.composer, source.creationTimestamp,
            source.discNumber, source.explicitType, source.inLibrary, source.isDeleted,
            source.lastRatingChangeTimestamp, source.playCount, source.recentTimestamp,
            source.totalDiscCount, source.totalTrackCount, source.estimatedSize ?: -1,
            source.trackAvailableForPurchase, source.trackAvailableForSubscription, source.trackType
    )

    override fun syncable(): SkyjamTrack {
        TODO("not implemented")
    }

}